numba/cuda/tests/cudapy/test_vectorize.py

```diff
@@ -227,6 +228,36 @@ def bar(x):
 
         self.assertEqual(bar.__name__, 'bar')
 
+    def test_no_transfer_for_device_data(self):
+        noise = np.random.randn(1, 3, 64, 64).astype(np.float32)
+        noise = cuda.to_device(noise)
+
+        # A mock of a CUDA function that always raises a CudaAPIError
+        def raising_transfer(*args, **kwargs):
+            raise CudaAPIError(999, 'Transfer not allowed')
+
+        old_HtoD = getattr(driver, 'cuMemcpyHtoD', None)
+        old_DtoH = getattr(driver, 'cuMemcpyDtoH', None)
+
+        setattr(driver, 'cuMemcpyHtoD', raising_transfer)
+        setattr(driver, 'cuMemcpyDtoH', raising_transfer)
+
+        try:
+            @vectorize(['float32(float32)'], target='cuda')
+            def func(noise):
+                return noise + 1.0
+
+            func(noise)
+        finally:
```

Comment:
I'll suggest adding a case where `raising_transfer` is called and the `CudaAPIError` is raised so we know the monkeypatch is working as expected.

Reply:
Good point!